Source/Core/Core/HW/SI/SI_DeviceGBA.cpp

```diff
@@ -233,140 +215,140 @@ void GBASockServer::ClockSync()
 }
 
 int GBASockServer::Receive(u8* si_buffer)
 {
   if (!m_client)
-    if (!GetAvailableSock(m_client))
-      return 5;
-
-  size_t num_received = 0;
-  u64 transfer_time = GetTransferTime((u8)m_send_data[0]);
-  bool block = (CoreTiming::GetTicks() - m_time_cmd_sent) > transfer_time;
-  if (m_cmd == CMD_STATUS && !m_booted)
-    block = false;
+    return 0;
 
-  if (block)
+  if (m_booted)
   {
     sf::SocketSelector selector;
     selector.add(*m_client);
     selector.wait(sf::milliseconds(1000));
   }
 
+  size_t num_received = 0;
+  std::array<u8, RECV_MAX_SIZE> recv_data;
   sf::Socket::Status recv_stat =
-      m_client->receive(m_recv_data.data(), m_recv_data.size(), num_received);
+      m_client->receive(recv_data.data(), recv_data.size(), num_received);
   if (recv_stat == sf::Socket::Disconnected)
   {
     Disconnect();
-    return 5;
+    return 0;
   }
 
   if (recv_stat == sf::Socket::NotReady)
     num_received = 0;
 
-  if (num_received > m_recv_data.size())
-    num_received = m_recv_data.size();
-
   if (num_received > 0)
   {
-#ifdef _DEBUG
-    if ((u8)m_send_data[0] == 0x00 || (u8)m_send_data[0] == 0xff)
-    {
-      WARN_LOG(SERIALINTERFACE, "%01d                              [< %02x%02x%02x%02x%02x] (%zu)",
-               m_device_number, (u8)m_recv_data[0], (u8)m_recv_data[1], (u8)m_recv_data[2],
-               (u8)m_recv_data[3], (u8)m_recv_data[4], num_received);
-    }
-    else
-    {
-      ERROR_LOG(SERIALINTERFACE, "%01d                              [< %02x%02x%02x%02x%02x] (%zu)",
-                m_device_number, (u8)m_recv_data[0], (u8)m_recv_data[1], (u8)m_recv_data[2],
-                (u8)m_recv_data[3], (u8)m_recv_data[4], num_received);
-    }
-#endif
-
-    for (size_t i = 0; i < m_recv_data.size(); i++)
-      si_buffer[i ^ 3] = m_recv_data[i];
+    for (size_t i = 0; i < recv_data.size(); i++)
+      si_buffer[i ^ 3] = recv_data[i];
   }
-
-  return (int)num_received;
+  return static_cast<int>(std::min(num_received, recv_data.size()));
 }
 
-CSIDevice_GBA::CSIDevice_GBA(SIDevices device, int device_number)
-    : ISIDevice(device, device_number), GBASockServer(device_number)
+CSIDevice_GBA::CSIDevice_GBA(SIDevices device, int device_number) : ISIDevice(device, device_number)
 {
 }
 
-CSIDevice_GBA::~CSIDevice_GBA()
-{
-  GBASockServer::Disconnect();
-}
-
 int CSIDevice_GBA::RunBuffer(u8* buffer, int length)
 {
-  if (!m_waiting_for_response)
+  switch (m_state)
   {
-    for (size_t i = 0; i < m_send_data.size(); i++)
-      m_send_data[i] = buffer[i ^ 3];
-
-    m_num_data_received = 0;
-    ClockSync();
-    Send(buffer);
+  case ConnectionState::AcceptCommand:
+  {
+    m_sock_server.ClockSync();
+    if (m_sock_server.Connect())
+    {
+#ifdef _DEBUG
+      NOTICE_LOG(SERIALINTERFACE, "%01d cmd %02x [> %02x%02x%02x%02x]", m_device_number, buffer[3],
+                 buffer[2], buffer[1], buffer[0], buffer[7]);
+#endif
+      m_sock_server.Send(buffer);
+    }
```

Comment:
Yeah, I did notice that when messing around with Four Swords the other day :/ I'll add `SI_ERROR_NO_RESPONSE` here in another commit since it shouldn't really change anything, and will make the disconnected case more obvious than it is now.